### ISSUE TITLE
importproject

### ISSUE THREAD

**Issue by Sairony** [premake/premake-core#1001]
Already did a post for it ( #991 ), but I can't seem to re-open the issue. In any case I still believe this would be a great addition. I've done a stab at it at [here](https://github.com/Sairony/premake-core), it works but some tests doesn't pass, my understanding of lua isn't good enough for me to easily be able to fix it sadly. Currently it's only tested on MSVC ( needed to change one line in sln2005.projects() to make sure it computes the correct path ). There's a small example [here](https://github.com/Sairony/premake-core/blob/master/tests/test_importproject.lua) that roughly shows how we use it. I understand this take on it won't make it into the master branch, but might be interesting for others to have a look at.

**Comment by samsinsane** [comment]
Have you tried [`includeexternal`](https://github.com/premake/premake-core/wiki/includeexternal) it kind of seems like it's everything you'd want?

Also, you might want to try using [tokens](https://github.com/premake/premake-core/wiki/Tokens), you can change `targetdir( "bin/" .. prj.name )` to `targetdir "bin/%{prj.name}"` along with numerous others. I'd also recommend looking into [`filters`](https://github.com/premake/premake-core/wiki/Filters) they deprecate the `configuration` API and it makes it much more obvious what you're trying to match against.

**Comment by Sairony** [comment]
I stuck with my frankenstein implementation, but now looking to upgrade one year later. I should check out the tokens stuff for sure. 

Anyway, `includeexternal` doesn't seem to work that well, I just get a lot of empty solutions. I think it's because my premake.lua files contain workspaces already? Basically I have premake_master.lua in the root. It goes through all the folders in the project sub folder & check them for a premake.lua. If there's one, it invokes it with `dofile()`. Inside that premake.lua file there's workspace containers ( usually just 1 ) containing projects and importing other projects which they depend on. At the end premake_master.lua imports all projects found into a master workspace which can be invoked to build everything. So if I want it to work with `includeexternal` I guess would have to split all the projects out into their own lua files, create new lua files for the project specific workspaces which then references the project ones, and then when I want to import a project, I need to import that specific projects lua file?

Also there seems to be a bug in api.lua pertaining to `function includeexternal(fname)`. If one include an external lua file, which in turn includes another external lua file, it will set `api._isIncludingExternal = nil` before the top most include external completes. It should probably be something like:
```
	function includeexternal(fname)
		local fullPath = p.findProjectScript(fname)
		local prevIsExternal = api._isIncludingExternal
		api._isIncludingExternal = true
		fname = fullPath or fname
		dofile(fname)
		api._isIncludingExternal = prevIsExternal
	end
```
But that's just speculation on my part because I don't understand how it works yet.

**Comment by WorldofBay** [comment]
yes that actually looks like a bug there and your change looks reasonable.
you can open a pull request for that and i think it'll get merged fast.
